src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs

```diff
@@ -63,5 +78,39 @@ internal override async Task<IPackageSearchMetadata> GetPackageMetadataAsync<T>(
 
             return await base.GetPackageMetadataAsync(identity, includePrerelease, cancellationToken);
         }
+
+        internal static PackageCollectionItem GetLatestApplicableTransitivePackageVersion(IGrouping<string, PackageCollectionItem> groupedPackagesById, IEnumerable<PackageCollectionItem> installedPkgs)
+        {
+            if (groupedPackagesById == null || !groupedPackagesById.Any())
+            {
+                throw new ArgumentException(Strings.Argument_Cannot_Be_Null_Or_Empty, nameof(groupedPackagesById));
+            }
+            if (installedPkgs == null)
+            {
+                throw new ArgumentNullException(nameof(installedPkgs));
+            }
+            if (groupedPackagesById
+                    .Any(group => !group.PackageReferences.Any() // PackageReferences is empty
+                        || group.PackageReferences.Any(pkgRef =>
+                            pkgRef is not ITransitivePackageReferenceContextInfo tr // it is not a transitive package
+                            || !tr.TransitiveOrigins.Any()))) // No transitive origin for a transitive package
+            {
+                throw new ArgumentException(Strings.InvalidCollectionElementType, nameof(groupedPackagesById));
```

Comment:
I don't really like that we're enforcing the data type of the elements of the group and their respective PackageReferences  with a runtime exception. There are other ways of handling that the data type of the PackageReferences should be what you expect them to be. I would consider restructuring this to have this method expect a collection of the right type at compile time to avoid throwing the exception.